### ISSUE TITLE
Consider using peer dependencies for plugins

### ISSUE THREAD

**Issue by nfantone** [neostandard/neostandard#184]
I wonder why external plugins, such as `eslint-plugin-n` or `eslint-plugin-promise`, are not defined as `"peerDependencies"`?

Today, I tried upgrading my local `eslint-plugin-n` from `17.10.3` to `17.11.1` (a *minor* bump), which I'm using for my own purposes, and `eslint` wouldn't run.

```sh
Oops! Something went wrong! :(

ESLint: 9.12.0

ConfigError: Config "node/flat/recommended-module": Key "plugins": Cannot redefine plugin "n".
```

Turns out, it conflicts with `neostandard` own `eslint-plugin-n` version. This is precisely the use case peer dependencies are good at handling. IMHO, there's no need for re-exporting your own internal dependencies as `{ plugins }`. Those should be kept internal — and let folks use their own.

**Comment by mcollina** [comment]
I'm somewhat reluctant to use `peerDependencies`, but they seem a good use case for this.

**Comment by bcomnes** [comment]
npm installs them for you automatically now, so it seems like less of a dx issue than it used to be. Been running neostandard without pinning my own eslint on projects and it seems to resolve reliably from neostandard,

**Comment by nfantone** [comment]
> npm installs them for you automatically now, so it seems like less of a dx issue than it used to be.

☝🏼 This. `yarn` does as well.

I'm happy to be challenged, but I'm not sure I see any downsides of leveraging peer deps here.

**Comment by voxpelli** [comment]
One of the main purpose with flat configs (and with [`standard-engine`](https://github.com/standard/standard-engine) that drives [`standard`](https://github.com/standard/standard)) is that peer dependencies are no longer needed – instead the module can be self-contained.

`neostandard` is not pinning the `eslint-plugin-n` dependency: https://github.com/neostandard/neostandard/blob/65b89ba7ed1d545665e18fab7055b28393a59b84/package.json#L84

If you want a newer version of it you can run [`npm update`](https://docs.npmjs.com/cli/v10/commands/npm-update) to refresh your dependencies to the latest version or use one of the tricks to update the subdependencies of `neostandard` specifically (eg. `npm install neostandard` may work or eg. `npm uninstall neostandard && npm install -D neostandard` or something else – depending on what you use to install npm modules)

If the `eslint-plugin-n` would have been outside of the range allowed by `neostandard`, then you can still enforce it by setting [`"override"`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#overrides) in your `package.json` (or whatever similar instruction that `pnpm`, `yarn` or such supports).

(You could maybe also use `"override"` temporarily to force the lockfile to be updated even when `neostandard` supports the target version, and then remove it afterwards)

**Comment by jerome-benoit** [comment]
> One of the main purpose with flat configs (and with [`standard-engine`](https://github.com/standard/standard-engine) that drives [`standard`](https://github.com/standard/standard)) is that peer dependencies are no longer needed – instead the module can be self-contained.

I do not understand how eslint flat configuration make peer dependencies not needed. Is there a reference document about it? 

> If you want a newer version of it you can run [`npm update`](https://docs.npmjs.com/cli/v10/commands/npm-update) to refresh your dependencies to the latest version or use one of the tricks to update the subdependencies of `neostandard` specifically (eg. `npm install neostandard` may work or eg. `npm uninstall neostandard && npm install -D neostandard` or something else – depending on what you use to install npm modules)
> 
> If the `eslint-plugin-n` would have been outside of the range allowed by `neostandard`, then you can still enforce it by setting [`"override"`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#overrides) in your `package.json` (or whatever similar instruction that `pnpm`, `yarn` or such supports).

I think the fine grained usage and maintenance of neostandard deps should be documented. And that for example dependabot will not upgrade them (no similar feature as renovate's lock file maintenance).

**Comment by nfantone** [comment]
@voxpelli I'm not sure I follow. Have the exact same questions as @jerome-benoit. 

> `neostandard` is not pinning the `eslint-plugin-n` dependency.

It wouldn't be needed for the problems that peer dependencies solve to arise. I can assure you, updating `eslint-plugin-n` _can_ lead to `eslint` breaking due to `neostandard` constraints.

**Comment by voxpelli** [comment]
> I do not understand how eslint flat configuration make peer dependencies not needed. Is there a reference document about it?

Since it moves from `"plugins": ["n"]` to eg. `plugins: { n: require('eslint-plugin-n') }`, so the module is imported before the config is defined rather than afterwards by `eslint`, so any module that exports configs (like `neostandard`) can have those modules resolved from their own location rather than rely on `eslint` and as such don't need them to be available at the top.

> I think the fine grained usage and maintenance of neostandard deps should be documented. And that, for example, dependabot will not upgrade them (no similar feature as renovate's lock file maintenance).

Sure, but this is not really something unique to `neostandard`. Sub-dependencies of all modules should be kept up to date and its important to eg. run `npm update` from time to time to ensure that deduplication and hoisting gets reapplied, as with at least `npm` every new module and module version you install will typically use the latest version of all its dependencies rather than whatever older version that currently exists in your `node_modules`, meaning that if one only ever updates modules through `npm install neostandard@latest` or similar one will have a _lot_ of duplicated modules in ones `node_module` folder. (Also why its good to update multiple modules at once, like `npm install neostandard@latest eslint@latest`, as then they will be installed together and deduplication can then be handled better)

> It wouldn't be needed for the problems that peer dependencies solve to arise. I can assure you, updating `eslint-plugin-n` can lead to `eslint` breaking due to `neostandard` constraints.

The error you get `Cannot redefine plugin "n".` is because the instance of plugins that share the same name in ESLint needs to be exactly the same in every config, and `neostandard` resolves its modules relative to itself whereas if you define and import one from your own `package.json` that's another instance in all cases except where the hoisting and de-duplication of `npm` merges them.

That's why `neostandard` is [re-exporting the modules](https://github.com/neostandard/neostandard?tab=readme-ov-file#exported-plugins) so that in those cases where one needs to reference it (eg. one wants to use a recommended config from one of the modules) one can: `plugins.n.configs['flat/recommended']`.

Try this (or equivalent using `yarn`, `pnpm`):

1. `npm uninstall eslint-plugin-n neostandard`
2. `npm install -D neostandard`

Since you're installing the plugin from scratch it should pick the latest dependency version of every subdependency of `neostandard`, which [right now](https://www.npmjs.com/package/eslint-plugin-n) is [`17.11.1`](https://www.npmjs.com/package/eslint-plugin-n/v/17.11.1):

https://github.com/neostandard/neostandard/blob/65b89ba7ed1d545665e18fab7055b28393a59b84/package.json#L84

**Comment by nfantone** [comment]
@voxpelli I get all of that. But I believe the point still stands. What should be seamless for users (updating a dependency) becomes sort of a chore. Plus, it requires knowledge of `neostandard` internal to even get to th

**Comment by nfantone** [comment]
@voxpelli I get all of that. But I believe the point still stands.

What should be seamless for users (updating a dependency) becomes sort of a chore. Plus, it requires knowledge of `neostandard` internals to even get to the bottom of the error that you're bound to trip over eventually, given this circumstances.

I understand the need. However, the way I see it, the fact that there's even a need to re-export dependencies as `{ plugins }` is already looking like a code smell.

**Comment by voxpelli** [comment]
> What should be seamless for users (updating a dependency) becomes sort of a chore.

This is because the lockfiles goes to great length to _not_ update the dependencies of anything but the very thing you tell it to install.

If you do `npm install eslint-plugin-n` and your current `neostandard` is using an older version that's been hoisted – then eg. `npm` will _actively_ de-hoist that version of `eslint-plugin-n` into `neostandard/node_modules/` and only after that install the new one.

This is a _feature_ of the package manager, not a bug.

If you use `npm install eslint-plugin-n` to update a subdependency of `neostandard`, then you are giving npm the wrong instructions.

Also:

The namespace for a plugin can be set to whatever by its user. Setting it to eg. `n` for `eslint-plugin-n` is [just a convention](https://eslint.org/docs/latest/use/configure/plugins#configure-plugins).

This, by intent I believe, enables one to use multiple version of the same plugin simultaneously. Eg. a new version of `eslint-plugin-n` can be used side by side with the one that `neostandard` supplies if eg. a new rule is needed and the new version is not yet supported by `neostandard` (such as when a future `18.0.0` gets released)

```js
{
  plugins: {
    'n': ...,
    'n-modern': ...,
  }
}
```

If we change to peer dependencies then you will not be able to use any `neostandard` incompatible version of `eslint-plugin-n` until `neostandard` itself support it (or you decide to ignore the peer dependency warning and use a version that's explicitly unsupported)

**Comment by bcomnes** [comment]
Trying to parse out the issue from the root cause:

- New incorrect rules went out via a transitive plugin update, that was later reverted upstream in eslint-plugin-n. A fix is quickly identified and released in range. https://github.com/neostandard/neostandard/issues/183
- People try to update to it as if it were a peerDependency, by installing  eslint-plugin-n top level next to neostandard without re-resolving a de-duplicated dep tree. They could have just run `npm update` to get this fix.
- Because the dependencies are not peer deps, trying to update them as if they are didn't work as expected. 
- Two distinct copies of eslint-plugin-n are now in the dep tree due to lockfile mutation without re-resolving the tree.
- The second instance of eslint-plugin-n is loaded into eslint for additional customizations and eslint throws the namespace error because the instances are distinct within the same namepace.

```
ConfigError: Config "node/flat/recommended-module": Key "plugins": Cannot redefine plugin "n".
```

What would some solutions look like for this issue:

- Use PeerDependencies: opt for the Dimond dependency problem so people can steer neostandard internal deps from parent package.jsons, but also loose the ability to install conflicting ranges of shared peers, and you also have to trust they resolve correctly near existing lockfiles. This is a non-trivial and sometimes unsolvable problem in some projects.
- Do nothing and document: remind users that its necessary to regenerate and update transitive dependencies periodically, and that they work differently than peerDependencies. Ensure its clear in the README how to namepace conflicting instances of plugins, and also how to avoid conflicting plugin instances with the public API re-exports of the plugins. 
- Name-spacing the neostandard registrations of its internally defined plugin instance. This would let poeple re-define duplicate plugins with the default namespace without issue, but makes customization of the neostandard namespaces more difficult. Massive breaking change, so probably not worth it. 

> What should be seamless for users (updating a dependency) becomes sort of a chore. 

Updating transitives is more seamless than updating peers (doesn't require agreeing range definitions), but it needs to be done. It wasn't done here. 

> Plus, it requires knowledge of `neostandard` internals to even get to the bottom of the error that you're bound to trip over eventually, given these circumstances.

neostandard is re-exporting the instance of a plugin it used in its public API, so that you can make subsequent use of of the guaranteed same instance. By being apart of the public API, this is by definition not a neostandard internal. Requiring matching instances in a namespace is an eslint limitation.

I'm totally on the fence here. PeerDeps just seems like we are buying wiggle room for avoiding transitive management, but also so many projects simply can't manage to keep on top of this these days. Eslint 9 is supposed to allow for the avoidance of peerDeps, but also the limitation around matching instances seems like it still is a soft requirement for it. 

@mcollina any elaboration on why you think its appropriate?

**Comment by voxpelli** [comment]
On the very topic of `eslint-plugin-n`, did a maintenance release now that bumped almost all dependencies: [`0.11.7`](https://github.com/neostandard/neostandard/releases/tag/v0.11.7)

**Comment by whaaaley** [comment]
+1 for peer deps. It's what, for the most part, eslint configs have landed on as the best option. The amount of headache I had to go through before landing here is silly. The whole idea of using a `standard-js` config is to not have to do anything. Set it, forget it. No one is going be mad at not being able to up date their promise linting package, or import linting package for a handful of rules they problem don't even really understand. Most people using `eslint-config-standard` probably don't even know they exist. Not having to understand is the whole idea.

**Comment by voxpelli** [comment]
> It's what, for the most part, eslint configs have landed on as the best option.

This is not really an accurate representation of it I think. Peer dependencies was what the legacy format required, since there the plugins were loaded by `eslint`, not the config.

Most reusable eslint configs still support the legacy format, very few has moved to become flat configs only, and since they support the legacy format they have to add the plugins as peer dependencies.

So it's not so much about it being "the best option" as it is about it being the only option.

One reusable config that has moved to become flat config only is `eslint-config-xo`and `eslint-config-xo-typescript`, and as one can see in https://github.com/xojs/eslint-config-xo-typescript/commit/5c79f4637b0c11488df2f9935b26f461dd2c4f44 they also moved away from peer dependencies, and `xo` at large will move even more plugins to become dependencies: https://github.com/xojs/xo/issues/769

> The whole idea of using a `standard-js` config is to not have to do anything. Set it, forget it.

Exactly, hence why the plugins are dependencies, as they should be managed by `neostandard`, not by the user.

> I do want to add that not having eslint-plugin-n, eslint-plugin-import, and eslint-pluin-promise installed cause the vscode extension to silently fail. These deps are not documented anywhere at all and are absolutely required in this case.

I am using the [VSCode ESLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) all the time and with great success.

The only finicky part I have experienced is that it doesn't always pick up on changes made to ones config, especially not eg. picking up a changed gitignore or such. One can then easily force VSCode to refresh itself by running the `Developer: Reload Window` or possibly `Developer: Restart Extension Host` from the command palette

**Comment by jerome-benoit** [comment]
> > It's what, for the most part, eslint configs have landed on as the best option.
> 
> This is not really an accurate representation of it I think. Peer dependencies was what the legacy format required, since there the plugins were loaded by `eslint`, not the config.

And trying to make `neostandard()` works with peer dependencies is an implementation and maintenance nightmare: 
- control flow to check if a peer deps is here
- refactor flat configurations to be peer dependencies tangled 
- a lot control flow to enable the relevant flat configuration section or not, especially in the common settings. And they will be heavy if a configuration is needed by several peer deps    

Adding big fat warning to the relevant documentation sections about embedded deps usage to ensure people will not miss it will be enough.

**Comment by mcollina** [comment]
> @mcollina any elaboration on why you think peers would be appropriate?

@bcomnes I don't.

**Comment by voxpelli** [comment]
> > @mcollina any elaboration on why you think peers would be appropriate?
> 
> @bcomnes I don't.

Oh, I also interpreted [`but they seem a good use case for this`](https://github.com/neostandard/neostandard/issues/184#issuecomment-2414483778) like you meaning peer dependencies could be appropriate here

Since all of us maintainers then being against using peer dependencies for plugins as I hear it, I think we can close this?

**Comment by mcollina** [comment]
> Oh, I also interpreted https://github.com/neostandard/neostandard/issues/184#issuecomment-2414483778 like you meaning peer dependencies could be appropriate here

I meant that I didn't want to oppose it if the others wanted to go for it. I don't think peerDependencies have a good ROI on maintainer time.
